esd_services_api_client/crystal/_connector.py

```diff
@@ -280,3 +287,47 @@ def dispose(self) -> None:
         Gracefully dispose object.
         """
         self.http.close()
+
+    def await_runs(
+        self, algorithm: str, run_ids: Optional[List[str]] = None, tags: Optional[List[str]] = None
+    ) -> Iterator[RequestResult]:
+        """
+        Await for a list of submitted Crystal jobs to complete.
+
+        :param algorithm: Name of an algorithm.
+        :param run_ids: Request identifiers assigned to the jobs by Crystal.
+        :param tags: Request tags assigned to the jobs by a client.
+        """
+        return self._await_runs(algorithm=algorithm, run_ids=run_ids, tags=tags)
+
+    def _await_runs(self, algorithm: str, run_ids: Optional[List[str]] = None, tags: Optional[List[str]] = None) -> Iterator[RequestResult]:
+        if tags is None:
+            tags = []
+
+        if run_ids is None:
+            run_ids = []
+
+        finished_statuses = [
+            RequestLifeCycleStage.COMPLETED,
+            RequestLifeCycleStage.FAILED,
+            RequestLifeCycleStage.SCHEDULING_TIMEOUT,
+            RequestLifeCycleStage.DEADLINE_EXCEEDED
+        ]
+
+        while len(tags) > 0:
+            for tag in tags[:]:
+                results = self._retrieve_runs(tag=tag, algorithm=algorithm)
+                if all(result.status in finished_statuses for result in results):
+                    tags.remove(tag)
+                    yield from results
+            if len(tags) > 0:
+                doze(1)
+
+        while len(run_ids) > 0:
```

Comment:
if you do this in a thread, you won't need for loop or while, i think

Reply:
Threaded but still looped